src/lib/registration/ui/migration_repos_workflow.rb

```diff
@@ -618,38 +618,91 @@ def prepare_media_upgrade
+      end
+
+      #
+      # Scan the installation repository and return the found product directories.
+      #
+      # @return [Array<Y2Packager::ProductLocation>] the found product directories
+      def full_medium_products
+        # in offline upgrade add the repository with the selected base product
+        # FIXME: similar to clients/inst_complex_welcome.rb and widgets/product_selector.rb
+        url = Yast::InstURL.installInf2Url("")
+        products = Y2Packager::ProductLocation
+                   .scan(url)
+                   .select { |p| p.details && p.details.base }
+        log.info("Found base products on the offline medium: #{products.pretty_inspect}")
+        products
+      end
+
+      #
+      # Find the upgrade base product from the Full medium for the installed
+      # base product.
+      #
+      # @param installed_base [<Y2Packager::Resolvable>] the installed base product
+      # @return [Y2Packager::ProductLocation, nil] The location of the new base product on
+      #  the Full installation medium, nil if no suitable upgrade product is found
+      def full_medium_upgrade_product(installed_base)
+        new_base = nil
+        base_products = full_medium_products
+        installed_names = Y2Packager::Resolvable.find(kind: :product, status: :installed)
+                                                .map(&:name)
+        # first check if there is a product rename defined for the installed products
+        # and the new renamed base product is available
+        # key in the mapping: list of installed products, value: the new base product
+        Y2Packager::ProductUpgrade::MAPPING.each do |k, v|
+          if (k - installed_names).empty?
+            new_base = base_products.find { |p| p.details && p.details.product == v }
+          end
+        end
+
+        # if no product rename was found then use the 1:1 upgrade
+        new_base ||= base_products.find do |p|
+          p.details && p.details.product == installed_base.name
+        end
+
+        new_base
+      end
+
+      #
+      # Initialize the base product repository and select the base product
+      # to install.
+      #
+      # @param new_product [String] Name of the product to install
+      # @param dir [String] Directory on the medium with the selected base product repository
+      def add_full_medium_repository(new_product, dir)
         # FIXME: this is the same as in installation/widgets/product_selector.rb and other places
         show_popup = true
+        url = Yast::InstURL.installInf2Url("")
         log_url = Yast::URL.HidePassword(url)
-        Yast::Packages.Initialize_StageInitial(show_popup, url, log_url, new_base.dir)
+        Yast::Packages.Initialize_StageInitial(show_popup, url, log_url, dir)
 
         # select the product to install
-        Yast::Pkg.ResolvableInstall(new_base.details && new_base.details.product, :product, "")
+        Yast::Pkg.ResolvableInstall(new_product, :product, "") if new_product
```

Comment:
what happens here when nil is passed? Previously you call `new_product = new_base.details && new_base.details.product` so code expect also nil there.

Reply:
ah, there is condition. nevermind